Add pattern matching pin support for instance/class/global variables

jeremyevans opened the pull request:
Pin matching for local variables and constants is already supported,
and it is fairly simple to add support for these variable types.

Note that pin matching for method calls is still not supported
without wrapping in parentheses (pin expressions).  I think that's
for the best as method calls are far more complex (arguments/blocks).

Fixes [Bug #17724]

k-tsj commented:
@jeremyevans 
Thanks for your PR.

It would be nice if you could also update doc/syntax/pattern_matching.rdoc.

jeremyevans commented:
@k-tsj I added documentation for the instance/class/global variable pinning.  I noticed there wasn't an example for expression pinning (even though it was listed in the grammar in the appendix), so I added an example for that.
